dbstructure/query.sql

```diff
@@ -0,0 +1,507 @@
+                coalesce(groups[6], '') as typ,
+                args_with_default[2] is not null as has_default
+          ) as parsed
+        where parsed.inout <> 'OUT '
+      ), array[]::json[]) as proc_args
+    from
+      pg_proc p
+      join pg_namespace pn on pn.oid = p.pronamespace
+      join pg_type t on t.oid = p.prorettype
+      join pg_namespace tn on tn.oid = t.typnamespace
+      left join pg_class comp on comp.oid = t.typrelid
+      left join pg_description as d on d.objoid = p.oid
+    where
+      pn.nspname = any ($1)
+  ),
+
+
+  -- Tables
+
+  tables as (
+    select
+      c.oid as table_oid,
+      n.nspname as table_schema,
+      c.relname as table_name,
+      d.description as table_description,
+      (
+        c.relkind in ('r', 'v', 'f')
+
+        and pg_relation_is_updatable(c.oid, true) & 8 = 8
+
+        -- The function `pg_relation_is_updateable` returns a bitmask where 8
+        -- corresponds to `1 << CMD_INSERT` in the PostgreSQL source code, i.e.
+        -- it's possible to insert into the relation.
+        or exists (
+          select 1
+          from pg_trigger
+          where
+            pg_trigger.tgrelid = c.oid
+            and (pg_trigger.tgtype::integer & 69) = 69
+            -- The trigger type `tgtype` is a bitmask where 69 corresponds to
+            -- TRIGGER_TYPE_ROW + TRIGGER_TYPE_INSTEAD + TRIGGER_TYPE_INSERT
+            -- in the PostgreSQL source code.
+        )
+      ) as table_insertable,
+      (
+        pg_has_role(c.relowner, 'USAGE')
+        or has_table_privilege(c.oid, 'SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, TRIGGER')
+        or has_any_column_privilege(c.oid, 'SELECT, INSERT, UPDATE, REFERENCES')
+      ) as table_is_accessible
+    from
+      pg_class c
+      join pg_namespace n
+        on n.oid = c.relnamespace
+      left join pg_catalog.pg_description as d
+        on d.objoid = c.oid and d.objsubid = 0
+    where
+      c.relkind in ('v','r','m','f')
+      and n.nspname not in ('pg_catalog'::name, 'information_schema'::name)
+  ),
+
+
+  -- Source columns, query explanation at
+  -- https://gist.github.com/steve-chavez/7ee0e6590cddafb532e5f00c46275569
+
+  view_col_rels as (
+    select
+      c.oid as view_oid,
+      targetentries.attnum as view_col_position,
+      targetentries.table_oid,
+      targetentries.table_col_position
+    from
+      pg_class c
+      join pg_namespace n on n.oid = c.relnamespace
+      join pg_rewrite r on r.ev_class = c.oid
+      join pg_attribute a on a.attrelid = c.oid
+      , lateral (
+          select
+            matches[1]::smallint as attnum,
+            matches[2] as attname,
+            matches[3]::oid as table_oid,
+            matches[3]::smallint as table_col_position
+          from
+            regexp_matches(
+              r.ev_action,
+              ':resno (\d+) :resname (\S+) :ressortgroupref \d+ :resorigtbl (\d+) :resorigcol (\d+)',
+              'g'
```

Comment:
This looks like it really simplifies my work on https://gist.github.com/steve-chavez/7ee0e6590cddafb532e5f00c46275569.

I've noticed you don't do the cleaning work I did on the previous query. Does this only target the VIEW root select(not the subselects or SELECT in WHERE or other parts)?

Reply:
Amazing work on this by the way, figuring out how to parse ev_action in sql is a feat!

This is an experiment to simplify the query, not even sure if it's faster yet... your version is already very optimal. The idea here is to make regex simpler (and hopefully faster). But as you note, this skips the more precise cleaning from your version. To compensate, the targetentries are filtered by which ones are valid columns of the view with the same attnum and name. So this would always find all relevant columns, but it might also include some additional ones in some edge cases (e.g. calculated columns). This could only be worth it if it was radically faster, bit that doesn't seem to be the case.